### PR TITLE
Christina Wusinich Project

### PR DESCRIPTION
**Pull request by cwusinich**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,92 @@
...
+-	Behavior data: a .txt output from the MID task that indicates button presses, timing of stimuli, etc.
+-	MEG data: a CTF (this is the MEG system we have at the NIH) file of MEG data from the task
+-	MRI data: structural MRIs for use with source localization (I already have a script for converting the files from the DICOM site to .nii files from my lab)
+
+# Processing steps:  
+I am not sure that is the best way to do this, but this is what needs to happen to get the data from its raw state to stats that I can report. The bolded script names are the ones I would like to create for my package based off a naming convention I’ve seen in our fMRI processing pipeline, though they are still subject to change:
+
+## Manual steps before scripts:
+1.	Transfer MEG, behavior, and MRI data to data processing folder on biowulf (this is just done with a BASH command)
+2.	Update demographics spreadsheet for use in analysis
```
Comment (leej3) [inline README.md]: These may not require manual intervention. The more you can automate the more scalable and less error prone your processing will be.

File: README.md [hunk 2]
```diff
@@ -1,3 +1,92 @@
...
+3.	Set fiducial markers on MRI copy
+
+##	MEG preprocessing step 1
+1.	Script name will be **01_megpreprocess1**, and I would like it to do the following:
+-	 Change file name using CTF command (to make file names look like “subject#_MID.ds” so they are more uniform)
+-	Set markers for stimuli (cue markers) and output file (.txt file) of the times of the cue marks (this file will be used in the next step)
+-	I already have some of the CTF commands for this, but I would like to figure out how to give this script (and all of the following ones) a list of subject names to loop through so I can do this all at once.
```
Comment (leej3) [inline README.md]: snakemake may be a useful approach for this. Or since you are working in human neuroimaging using nipype might be worth it. These tools for building and executing workflows are hard and the beginning but they pay off in the long run. An added advantage is that nipype is that many of the generic tools for neuroimaging processing use it so you will be able to debug/modify [these pipelines](https://bids-apps.neuroimaging.io/apps/) more easily

File: README.md [hunk 3]
```diff
@@ -1,3 +1,92 @@
...
+2.	Input: MEG data file
+3.	Output: copied and edited MEG data file, cue markers text file
+
+##	Behavior processing
+1.	Script name will be **02_behprocess**, and it will do the following:
+-	Makes marker files for more MEG processing; these markers will designate win/loss/neutral cues to be marked in the MEG file and are output as three separate .txt files to the subject’s MEG directory
+-	Pulls columns from behavior data file and calculates mean RTs and accuracy by trial and subject and appends that to master behavior data sheet (I will probably work with this sheet in R unless I can do an ANOVA as another step)
```
Comment (leej3) [inline README.md]: Look into pandas for working on tabular data. It will make things a lot easier. And using csv files instead of text might be useful.

Comment (leej3) [reply README.md]: Python has advance tools for statistical learning/analysis. Statsmodels is probably what you need here but as the sophistication of your analysis grows you could add in the packages lme, scikit-learn, and tensorflow. There are certainly times to use R but if you can complete everything with python it will help you a lot. Learning how to debug bash, python, and R is never a fun task.

File: README.md [hunk 4]
```diff
@@ -1,3 +1,92 @@
...
+-	I have already made two separate jupyter notebook files to do each of the above tasks, but I can figure how to do it without finding and replacing the subject ID; obviously a loop would be a lot easier, so I would like to figure that out and collapse both notebooks into one script
```
Comment (leej3) [inline README.md]: I suspect a pandas dataframe, string formatting should be adequate. Your processing should likely be encapsulated in functions (and in turn in modules). Any function could take subject id as an argument to modify it's behavior appropriately.

File: README.md [hunk 5]
```diff
@@ -1,3 +1,92 @@
...
+##	MEG processing with SAM
+1.	Script name will be **04_sam** and will do the following:
+-	orthohull (from subject’s MRI folder)
+-	Source localization in high gamma using three SAM commands and parameter file→ output is .nii file
+-	Copy .nii file from subject’s MEG directory to their MRI directory
+-	Talairach SAM results for analysis in afni
+-	Copy talairached .nii file into a group folder for group analysis
+2.	Problems:
+-	This step uses a parameter file that I would like to be able to create and distribute among subject directories whenever we want to change parameters (e.g. the time window we’re looking at in the task)
+-	Each of the three SAM commands takes a bit of time, but I heard that using swarm in biowulf may be a good way to do this. I am not sure how it would work with a loop just yet, but it’s an aspiration to make this somewhat efficient if possible.
```
Comment (leej3) [inline README.md]: This is where nipype would shine. It is true that swarm would distribute this but then it is not very portable. And workflow managers do lots of intelligent management of the processing so that when you make a change you only recompute the bits of the analysis that you need to.

File: README.md [hunk 6]
```diff
@@ -1,3 +1,92 @@
...
+##	Analysis
+1.	Right now, I will plan to make this a step for 3dLME in afni, but we may choose to do a different analysis in the coming weeks.
+2.	Script name will be **05_3dLME** and will do the following:
+-	Runs 3dLME on all the .nii files
+-	I already have a tcsh script that does this, so I’m not sure if this is really necessary to make it in python?
+3.	Input: do_3dLME_MID_highgamma2.txt (this is the text file I mentioned above that contains file paths of the .nii files and demographic info of participants, such as group and age)
+4.	Output: text file named by date and type of analysis with results from 3dLME
+
+##	Permissions
+1.	I want a script that will set all of the permissions correctly for my group for all of the output files above because umask in my bash profile does not seem to work for this.
```
Comment (leej3) [inline README.md]: Did you set your default group to one that other people are in? This surprises me.

File: README.md [hunk 7]
```diff
@@ -1,3 +1,92 @@
...
+2.	The commands in this script may just be put at the end of each of the above ones depending on how this works out, or should it just be a bash script?
```
Comment (leej3) [inline README.md]: Bash is fine. Python would be better. There's not too much point in rewriting lots of bash scripts into python. With that said if you tie everything together with python it will be easier to maintain and debug

File: README.md [hunk 8]
```diff
@@ -1,3 +1,92 @@
...
+
+# Additional Questions:  
+-	Can I have a script do “module load afni” (and ctf, and R) so that I don’t have to?
```
Comment (leej3) [inline README.md]: You can certainly put this in your ~/.bashrc file (or ~/.bash_profile, depends on what os you are on). For the class project you are developing a python package. A python package (a pip installed one) will often just assume the appropriate dependencies are installed on the system.

File: README.md [hunk 9]
```diff
@@ -1,3 +1,92 @@
...
+-	If the commands are from CTF, can they be in a python script? I had just been running them in the terminal, so would I have to add something special to the beginning of each one to include it in a python script?
```
Comment (leej3) [inline README.md]: The subprocess package may be what you need here. Or the above mentioned workflow managers

File: README.md [hunk 10]
```diff
@@ -1,3 +1,92 @@
...
+-	Can I make a giant script that will run all of these steps (01-05 and setting permissions) on a list of subjects?
```
Comment (leej3) [inline README.md]: It should be a giant script. It should be a script that provides a command line interface to the user (use arg parse) and then calls the appropriate functionality (imported from other python modules in your package).

Or you might just have several smaller scripts if that is more convenient. Python entrypoints might be useful as you attempt to achieve this. Using the scripts keyword for setup.py may also be useful if you do wish to install non-python scripts as part of your package.

File: README.md [hunk 11]
```diff
@@ -1,3 +1,92 @@
...
+-	How can I set my default python version in biowulf without manually having to load 3.7 every time? Can I do that at the beginning of the first script or in my profile?
```
Comment (leej3) [inline README.md]: yes. I  have a Conda environment that is set in my profile as a default. That way I will always be using the python in that environment.

Often `module load` can be quite slow. Often, I create an bash alias to software within a conda environment. so for example `alias git=/home/$USER/anaconda3/envs/misc_env/bin/git` takes no time to load at login but will give me git from that environment when I want it.

**Review by leej3 — APPROVED** [review]

File: README.md (hunk 2 above)
Comment (cwusinich) [reply README.md]: Thank you so much for all of these responses and suggestions!!

**Comment by leej3** [comment]
> Am I doing git right?

Yes it appears so

> Is there a reason that something would work when I run a command individually vs. running as a swarm? Trouble-shooting ideas would be great because my whole project won’t work if I can’t figure this out…

Sounds like checking both the .o and .e files will help to debug this.

> Related to q#1: how do I figure out how many CPUs and threads to assign to a job? I’ve been doing a lot of googling and am still a bit confused. I’ve just been assigning a lot until the command runs successfully, but that seems kind of sloppy and maybe a disrespectful use of shared resources. And what about threads?

No need to worry about taking more than you have allocated, they don't let you. You only harm your own processing. It is an sbatch system so likely you can use variables like the ones listed [here](https://wiki.hpcc.msu.edu/display/ITH/SLURM+Environment+Variables) Try: SLURM_CPUS_PER_TASK.

> I need some help conceptualizing how to bring multiple scripts together so that I can run them in order on a bunch of subjects at once while allocating enough memory for each so that this doesn’t take forever. Also I have a combination of python scripts I’ve written and some bash swarm files (which one of the python scripts generates), which is adding to some of my confusion. In Dr. Lee’s comments on my outline, he mentioned looking into pipelines, which I hope to do by next week but will probably still have some questions about organizing this and which one to pick.

Have a look at snakmake and nipype. They're worth it for the long run. Just maybe not for this project

Other questions if there is time:
Additionally, I have been encountering a lot of errors when swarming a bunch of subjects at once which sometimes seem to get lost in the shuffle (so I don’t realize it until a later step) or are very vague (i.e. it will just say “FAILED” and I have to then go figure out why); so it would be nice to figure out how to build in some more informative error messages and mechanisms that will stop later scripts in the chain if one of the earlier ones fails (also messages that tell me if things have run successfully--maybe some kind of report output?).
Another higher level question is that the neuroimaging data I’m working with just seems to be kind of unpredictable (or error-inducing?), and I wonder if it’s even wise to combine too many steps into one, though maybe the error messages will help with that. Or maybe is there something I can do early in the process to make sure the datasets are ready to go through the pipeline?

File: README.md (hunk 7 above)
Comment (leej3) [reply README.md]: It occurred to me that some tools completely ignore umask. Perhaps that is your problem. Assuming you have written all your output to "/data/$USER/output_dir" you could make sure your output is well behaved regarding permissions by ending each swarm job with something like:

```
OUTPUT_DIR=/data/$USER/output_dir
GROUP=mygroup
find $OUTPUT_DIR -group $USER -exec chown :$GROUP {} \; -exec chmod g+rx \;
```

**Comment by leej3** [comment]
f-strings will make a lot of this easier to construct and more legible. For example...

```
subject = "01"
session = "a"
cmd = f"""
    command 
        -ax 
        --another-flag 
        -s {subject}
        --session {session}
    """
cmd = ' '.join(cmd.split())
print(cmd)
```

Will generate "command -ax --another-flag -s 01 --session a". It is a lot more legible than adding strings together. Something to bear in mind in future.

Also, add a setup.py and some tests...

**Comment by cwusinich** [comment]
Thank you for the f-strings suggestion! If I don't change it for the project deadline, I definitely will at some point moving forward.

**Comment by cwusinich** [comment]
@leej3 If we don't have any functions (I just have scripts), what should the tests be? Should I make one of them a function just so it can have tests?

**Comment by leej3** [comment]
Yes, check the rubric. If you don't have functions it is hard to test, and
your code will not be modular so it will be difficult to
understand/maintain/reuse.

**Comment by cwusinich** [comment]
Thanks! Will do.

**Comment by cwusinich** [comment]
@leej3 I'm still planning to add more tests today and clean some of the scripts up if there's time, but the one test I have in there now works, and so do all of my modules, as far as I can tell. I also tested installing the package on a different computer, and that worked, and the test ran successfully. I'm wondering why it's not passing circleci; I read the details of it failing and looked at other people's pull requests that passed, but I'm still a little confused about what's wrong with mine. I'm guessing something isn't formatted or set-up correctly, but I'm not sure what. If you have time, can you check to see what's going wrong?

**Comment by leej3** [comment]
Great work. I've submitted some suggested fixes to help you on your way.

Quick summary:
+ Encapsulating fake data creation is a good idea
+ test modules must start with "test" for pytest to pick them up
+ Try to use pathlib.Path more. Just as f-string change one's life with strings, Path objects make working with paths much easier. The one gotcha (which is becoming less frequent) is that some tools insist on a string as input instead of a Path object. The quick fix is `my_path_as_a_string = str(my_path_object)`

**Comment by cwusinich** [comment]
Thank you! It looks so much better now!

**Comment by cwusinich** [comment]
Submitting now. Thank you for your help with this! I learned _way_ more than I thought I would!

**Comment by leesup** [comment]
You did an awesome job on your project! You provided truly impressive description of your project and tests. Great job using Numpy and pandas - I hope these become your best friends when using Python! I would suggest having more white space for improved readability. In addition, I would suggest checking whether the libraries/modules you imported are actually being used. I saw that you imported csv module, but you ended up using pandas csv instead. Using pandas is completely fine, but you don’t need to import csv in that case. Also, you should have a look into pathlib and argparse. Again, great job with your project!

-Paul

**Comment by cwusinich** [comment]
Thank you @suhwanplee !! Those are great suggestions; I am definitely going to look into them. Take care!
